src/Stratis.Bitcoin/AddressIndexing/AddressIndexer.cs

```diff
@@ -0,0 +1,284 @@
+
+                this.logger.LogDebug("TxIndexing is enabled.");
+
+                this.dataStore = this.db.GetCollection<AddressIndexerData>(DbKey);
+
+                this.data = this.dataStore.FindAll().FirstOrDefault();
+
+                if (this.data == null)
+                {
+                    this.logger.LogDebug("Tip was not found, initializing with genesis.");
+
+                    this.data = new AddressIndexerData()
+                    {
+                        TipHash = this.network.GenesisHash.ToString(),
+                        AddressIndexDatas = new List<AddressIndexData>()
+                    };
+                    this.dataStore.Insert(this.data);
+                }
+
+                this.blockReceivedQueue = new AsyncQueue<KeyValuePair<bool, ChainedHeaderBlock>>(this.OnEnqueueAsync);
+
+                // Subscribe to events.
+                this.blockConnectedSubscription = this.signals.Subscribe<BlockConnected>(blockConnectedData =>
+                {
+                    while (this.blockReceivedQueue.Count > 100)
+                    {
+                        this.logger.LogWarning("Address indexing is slowing down the consensus.");
+                        Thread.Sleep(5000);
+                    }
+
+                    this.blockReceivedQueue.Enqueue(new KeyValuePair<bool, ChainedHeaderBlock>(true, blockConnectedData.ConnectedBlock));
+                });
+
+                this.blockDisconnectedSubscription = this.signals.Subscribe<BlockDisconnected>(blockDisconnectedData =>
+                {
+                    this.blockReceivedQueue.Enqueue(new KeyValuePair<bool, ChainedHeaderBlock>(false, blockDisconnectedData.DisconnectedBlock));
+                });
+
+                this.nodeStats.RegisterStats(this.AddComponentStats, StatsType.Component, 400);
+            }
+        }
+
+        private Task OnEnqueueAsync(KeyValuePair<bool, ChainedHeaderBlock> item, CancellationToken cancellationtoken)
+        {
+            try
+            {
+                this.ProcessBlockAddedOrRemoved(item);
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogError(ex.ToString());
+                throw;
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private void AddComponentStats(StringBuilder benchLog)
+        {
+            benchLog.AppendLine();
+            benchLog.AppendLine("======AddressIndexer======");
+
+            benchLog.AppendLine($"Unprocessed blocks: {this.blockReceivedQueue.Count}");
+        }
+
+        /// <summary>Processes block that was added or removed from consensus chain.</summary>
+        private void ProcessBlockAddedOrRemoved(KeyValuePair<bool, ChainedHeaderBlock> item)
+        {
+            // Make sure it's on top of the tip.
+            bool blockAdded = item.Key;
+
+            if ((blockAdded && item.Value.ChainedHeader.Header.HashPrevBlock.ToString() != this.data.TipHash) ||
```

Comment:
`.ToString()` ouch why?

Reply:
Because serializer that litedb uses can't work with uint256

Reply:
:( won't it be then better to use `byte[]`?

Reply:
Fair point. Will do it in the next PR